### ISSUE TITLE
验证数据集时评价指标发生错误

### ISSUE THREAD

**Issue by Flacon12** [meituan/YOLOv6#844]
### Before Asking

- [X] I have read the [README](https://github.com/meituan/YOLOv6/blob/main/README.md) carefully. 我已经仔细阅读了README上的操作指引。

- [X] I want to train my custom dataset, and I have read the [tutorials for training your custom data](https://github.com/meituan/YOLOv6/blob/main/docs/Train_custom_data.md) carefully and organize my dataset correctly; (FYI: We recommand you to apply the config files of xx_finetune.py.) 我想训练自定义数据集，我已经仔细阅读了训练自定义数据的教程，以及按照正确的目录结构存放数据集。（FYI: 我们推荐使用xx_finetune.py等配置文件训练自定义数据集。）

- [X] I have pulled the latest code of main branch to run again and the problem still existed. 我已经拉取了主分支上最新的代码，重新运行之后，问题仍不能解决。

### Search before asking

- [X] I have searched the YOLOv6 [issues](https://github.com/meituan/YOLOv6/issues) and found no similar questions.

### Question

当我训练完获得权重时，在eval.py验证数据集后得到以下日志内容：
D:\anaconda\envs\hxs\python.exe D:/zzz/YOLOv6-0.4.0/tools/eval.py
Namespace(batch_size=32, conf_thres=0.03, config_file='', data='../data/voc.yaml', device='0,1', do_coco_metric=True, do_pr_metric=True, eval_config_file='./configs/experiment/eval_640_repro.py', half=False, img_size=640, infer_on_rect=True, iou_thres=0.65, name='exp', plot_confusion_matrix=True, plot_curve=True, reproduce_640_eval=False, save_dir='runs/val/', shrink_size=0, task='val', verbose=True, weights='D:\\zzz\\YOLOv6-0.4.0\\tools\\runs\\train\\exp60\\weights\\best_ckpt.pt')
Loading checkpoint from D:\zzz\YOLOv6-0.4.0\tools\runs\train\exp60\weights\best_ckpt.pt

Fusing model...
D:\anaconda\envs\hxs\lib\site-packages\torch\functional.py:504: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at C:\cb\pytorch_1000000000000\work\aten\src\ATen\native\TensorShape.cpp:3191.)
  return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]
Switch model to deploy modality.
Model Summary: Params: 18.54M, Gflops: 45.20
Val: Checking formats of labels with 8 process(es): 
381 label(s) found, 0 label(s) missing, 0 label(s) empty, 0 invalid label files: 100%|██████████| 381/381 [00:06<00:00, 56.79it/s] 
Convert to COCO format
100%|██████████| 381/381 [00:00<00:00, 47756.55it/s]
Convert to COCO format finished. Resutls saved in ..\annotations\instances_val.json
Val: Final numbers of valid images: 381/ labels: 381. 
7.0s for dataset initialization.
Inferencing model in val datasets.: 100%|███████| 12/12 [00:15<00:00,  1.28s/it]
IOU 50 best mF1 thershold near 0.602.
Class                 Images      Labels     P@.5iou     R@.5iou    F1@.5iou      mAP@.5  mAP@.5:.95
all                      381        1290        0.92       0.878       0.897       0.941       0.807
Black-grass              381          80       0.827       0.775         0.8       0.832       0.573
Charlock                 381         105       0.962       0.905       0.932       0.968       0.917
Cleavers                 381          66       0.942       0.879        0.91       0.949        0.88
Common Chickweed         381         170       0.972       0.924       0.947       0.976       0.816
Common wheat             381          70       0.895       0.853       0.873       0.892       0.699
Fat Hen                  381         115       0.945       0.852       0.896       0.912       0.736
Maize                    381          84       0.964        0.81        0.88       0.947       0.804
Scentless Mayweed         381         137       0.932       0.906       0.919       0.962       0.867
Shepherd Purse           381          58        0.86       0.897       0.878       0.967       0.883
Small-flowered Cranesbill         381         123           1       0.975       0.987       0.993       0.939
Sugar beet               381         130        0.83       0.938       0.881       0.945       0.876
Loose Silky-bent         381         152       0.905       0.829       0.865        0.95       0.686

Evaluating speed.
Average pre-process time: 0.31 ms
Average inference time: 4.18 ms
Average NMS time: 3.59 ms

Evaluating mAP by pycocotools.
Saving runs\val\exp10\predictions.json...
loading annotations into memory...
Done (t=0.03s)
creating index...
index created!
Loading and preparing results...
DONE (t=0.73s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=1.77s).
Accumulating evaluation results...
DONE (t=0.59s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.803
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.936
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.858
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.556
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.819
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.844
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.875
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.877
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.738
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.887
Class           Labeled_images      Labels     P@.5iou     R@.5iou    F1@.5iou      mAP@.5  mAP@.5:.95
all                      381        1290       0.914        0.86       0.886       0.936       0.803
Black-grass               80          80       0.813        0.76       0.786       0.816       0.568
Charlock                 105         105       0.969         0.9       0.933       0.964       0.913
Cleavers                  66          66       0.951        0.87       0.909       0.945       0.876
Common Chickweed         170         170       0.975        0.92       0.947       0.971       0.812
Common wheat              70          70       0.896        0.85       0.872       0.888       0.698
Fat Hen                  115         115       0.943        0.86         0.9       0.908       0.734
Maize                     84          84       0.971         0.8       0.877       0.944       0.801
Scentless Mayweed         137         137       0.933        0.91       0.921       0.958       0.863
Shepherd Purse            58          58       0.944        0.87       0.906       0.967       0.881
Small-flowered Cranesbill         123         123           1        0.97       0.985       0.989       0.933
Sugar beet               130         130       0.934        0.87       0.901        0.94       0.873
Loose Silky-bent         152         152       0.861        0.93       0.894       0.947       0.685
Results saved to runs\val\exp10

Process finished with exit code 0

但当我过几天再次在eval.py验证数据集时，评价指标都不对：
D:\anaconda\envs\hxs\python.exe D:/zzz/YOLOv6-0.4.0/tools/eval.py
Namespace(batch_size=32, conf_thres=0.03, config_file='', data='../data/voc.yaml', device='0,1', do_coco_metric=True, do_pr_metric=True, eval_config_file='./configs/experiment/eval_640_repro.py', half=False, img_size=640, infer_on_rect=True, iou_thres=0.65, name='exp', plot_confusion_matrix=True, plot_curve=True, reproduce_640_eval=False, save_dir='runs/val/', shrink_size=0, task='val', verbose=True, weights='D:\\zzz\\YOLOv6-0.4.0\\tools\\runs\\beifen\\CA\\P3P4simSPPF\\YU\\exp60\\weights\\best_ckpt.pt')
Loading checkpoint from D:\zzz\YOLOv6-0.4.0\tools\runs\beifen\CA\P3P4simSPPF\YU\exp60\weights\best_ckpt.pt

Fusing model...
D:\anaconda\envs\hxs\lib\site-packages\torch\functional.py:504: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at C:\cb\pytorch_1000000000000\work\aten\src\ATen\native\TensorShape.cpp:3191.)
  return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]
Switch model to deploy modality.
Model Summary: Params: 18.53M, Gflops: 45.20
Val: Checking formats of labels with 8 process(es): 
381 label(s) found, 0 label(s) missing, 0 label(s) empty, 0 invalid label files: 100%|██████████| 381/381 [00:05<00:00, 65.73it/s] 
Convert to COCO format
100%|██████████| 381/381 [00:00<00:00, 34729.96it/s]
Convert to COCO format finished. Resutls saved in ..\annotations\instances_val.json
Val: Final numbers of valid images: 381/ labels: 381. 
6.2s for dataset initialization.
Inferencing model in val datasets.: 100%|███████| 12/12 [00:41<00:00,  3.42s/it]
IOU 50 best mF1 thershold near 0.397.
Class                 Images      Labels     P@.5iou     R@.5iou    F1@.5iou      mAP@.5  mAP@.5:.95
all                      381        1290       0.283        0.11       0.141      0.0957      0.0726
Black-grass              381          80     0.00861       0.113       0.016      0.0517      0.0259
Charlock                 381         105       0.375       0.143       0.207       0.119      0.0853
Cleavers                 381          66       0.059       0.106      0.0758      0.0584      0.0493
Common Chickweed         381         170       0.116      0.0471       0.067      0.0171      0.0113
Common wheat             381          70         0.5      0.0857       0.146      0.0922      0.0638
Fat Hen                  381         115       0.202      0.0696       0.104      0.0459      0.0324
Maize                    381          84       0.605       0.202       0.303       0.227       0.182
Scentless Mayweed         381         137       0.147      0.0438      0.0675      0.0309      0.0264
Shepherd Purse           381          58      0.0448      0.0517       0.048     0.00871     0.00598
Small-flowered Cranesbill         381         123        0.73       0.171       0.277       0.283       0.244
Sugar beet               381         130       0.372       0.146       0.209       0.115      0.0857
Loose Silky-bent         381         152       0.239       0.138       0.175       0.101      0.0592

Evaluating speed.
Average pre-process time: 0.29 ms
Average inference time: 9.97 ms
Average NMS time: 25.48 ms

Evaluating mAP by pycocotools.
Saving runs\val\exp27\predictions.json...
loading annotations into memory...
Done (t=0.03s)
creating index...
index created!
Loading and preparing results...
DONE (t=1.71s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=3.47s).
Accumulating evaluation results...
DONE (t=1.57s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.076
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.099
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.087
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.019
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.092
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.141
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.193
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.211
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.053
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.223
Class           Labeled_images      Labels     P@.5iou     R@.5iou    F1@.5iou      mAP@.5  mAP@.5:.95
all                      381        1290       0.208        0.16       0.181      0.0993       0.076
Black-grass               80          80       0.121         0.1        0.11      0.0571      0.0305
Charlock                 105         105       0.362        0.16       0.222       0.123      0.0885
Cleavers                  66          66       0.444        0.06       0.106      0.0627       0.053
Common Chickweed         170         170      0.0918        0.05      0.0647      0.0194      0.0137
Common wheat              70          70       0.282        0.15       0.196      0.0979      0.0686
Fat Hen                  115         115       0.205        0.06      0.0928      0.0503      0.0366
Maize                     84          84        0.46        0.27        0.34       0.229       0.184
Scentless Mayweed         137         137       0.462        0.04      0.0736      0.0355      0.0311
Shepherd Purse            58          58      0.0612        0.05       0.055      0.0094     0.00656
Small-flowered Cranesbill         123         123       0.416        0.34       0.374       0.284       0.247
Sugar beet               130         130       0.548        0.13        0.21       0.119        0.09
Loose Silky-bent         152         152       0.408        0.13       0.197       0.105      0.0622
Results saved to runs\val\exp27

Process finished with exit code 0

**Comment by Flacon12** [comment]
eval.py文件我仅仅更改了权重文件路径和开启了这几个参数，其他参数都没有
    parser.add_argument('--do_coco_metric', default=True, type=boolean_string, help='whether to use pycocotool to metric, set False to close')
    parser.add_argument('--do_pr_metric', default=True, type=boolean_string, help='whether to calculate precision, recall and F1, n, set False to close')
    parser.add_argument('--plot_curve', default=True, type=boolean_string, help='whether to save plots in savedir when do pr metric, set False to close')
    parser.add_argument('--plot_confusion_matrix', default=True, action='store_true', help='whether to save confusion matrix plots when do pr metric, might cause no harm warning print')
    parser.add_argument('--verbose', default=True, action='store_true', help='whether to print metric on each class')

**Comment by mtjhl** [comment]
您可以 git diff 一下，看看是否有其他的改动。
